src/dataflow/src/render/reduce.rs

```diff
@@ -208,147 +304,318 @@ where
             // Distinct is a special case, as there are no aggregates to aggregate.
             // In this case, we use a special implementation that does not rely on
             // collating aggregates.
-            if aggregates.is_empty() {
-                let (oks, errs) = (
-                    ok_input.reduce_abelian::<_, OrdValSpine<_, _, _, _>>("DistinctBy", {
-                        |key, _input, output| {
-                            output.push((key.clone(), 1));
+
+            // First, let's plan out what we are going to do with this reduce
+            let lir = plan_reduce(aggregates.clone(), *monotonic, *expected_group_size);
+            let arrangement = build_reduce(ok_input, lir);
+            let index = (0..group_key.len()).collect::<Vec<_>>();
+            self.set_local_columns(
+                relation_expr,
+                &index[..],
+                (arrangement, err_input.arrange()),
+            );
+        }
+    }
+}
+
+// Generate a lower-level IR node for the supplied aggregations that
+// describes what the resulting dataflow will look like.
+fn plan_reduce(
+    aggregates: Vec<AggregateExpr>,
+    monotonic: bool,
+    expected_group_size: Option<usize>,
+) -> ReduceExpr {
+    let lower = move |typ: ReductionType, aggregates_list: Vec<(usize, AggregateExpr)>| {
```

Comment:
This really strikes me as something that might be better as a function. It could have a descriptive name, some documentation describing what it does, and more clearly explain what it closes over (I found `monotonic`; is it just that?).